apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/IncludeRequestDispatcherInstrumentation.java

```diff
@@ -0,0 +1,104 @@
+import co.elastic.apm.agent.impl.transaction.TraceContextHolder;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+import javax.servlet.RequestDispatcher;
+import javax.servlet.ServletRequest;
+import javax.servlet.http.HttpServletRequest;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static net.bytebuddy.matcher.ElementMatchers.*;
+
+public class IncludeRequestDispatcherInstrumentation extends ElasticApmInstrumentation {
+
+    private static final String SPAN_TYPE = "template";
+    private static final String SPAN_SUBTYPE = "servlet";
+    private static final String SPAN_ACTION = "include";
+
+    private static final String INCLUDE = "INCLUDE";
+
+    @Override
+    public Class<?> getAdviceClass() {
+        return IncludeRequestDispatcherAdvice.class;
+    }
+
+    public static class IncludeRequestDispatcherAdvice extends IncludeRequestDispatcherInstrumentation {
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        private static void beforeExecute(@Advice.Local("span") @Nullable Span span,
+                                          @Advice.Argument(0) @Nullable ServletRequest request) {
+            if (tracer == null || tracer.getActive() == null) {
+                return;
+            }
+            final TraceContextHolder<?> parent = tracer.getActive();
+
+            if (request != null && request instanceof HttpServletRequest) {
+                HttpServletRequest httpServletRequest = (HttpServletRequest) request;
+                String includeServletPath = (String) httpServletRequest.getAttribute(RequestDispatcher.INCLUDE_SERVLET_PATH);
+                span = parent.createSpan()
+                    .withType(SPAN_TYPE)
+                    .withSubtype(SPAN_SUBTYPE)
+                    .withAction(SPAN_ACTION)
+                    .withName(INCLUDE);
+                if (includeServletPath != null) {
+                    span.appendToName(" ").appendToName(includeServletPath);
+                }
+                span.activate();
+            }
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class)
+        public static void afterExecute(@Advice.Local("span") @Nullable Span span,
+                                        @Advice.Thrown @Nullable Throwable t) {
+
+            if (span != null) {
+                span.captureException(t)
+                    .deactivate()
+                    .end();
+            }
+        }
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return not(isInterface())
+            .and(nameContainsIgnoreCase("dispatcher"));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("include").and(takesArguments(2))
```

Comment:
Instead of duplicating the instrumentation, I'd rather match "include" and "forward" and inject the actual method name by adding a `@Advice.Origin("#m") String method` parameter to the `@OnMethodEnter` method.